src/coreclr/jit/lclmorph.cpp

```diff
@@ -965,6 +972,124 @@ class LocalAddressVisitor final : public GenTreeVisitor<LocalAddressVisitor>
     {
         assert(val.IsLocation());
 
+        ClassLayout*   indirLayout = nullptr;
+        IndirTransform transform   = SelectLocalIndirTransform(val, user, &indirLayout);
+
+        if (transform == IndirTransform::None)
+        {
+            return;
+        }
+
+        GenTree*      indir    = val.Node();
+        FieldSeqNode* fieldSeq = val.FieldSeq();
+
+        if (fieldSeq == nullptr)
+        {
+            fieldSeq = FieldSeqStore::NotAField();
+        }
+
+        if (fieldSeq != FieldSeqStore::NotAField())
+        {
+            // For non-FIELD access, we need to check whether the type of the last field in the
+            // sequence matches that of the indirection. For primitives, this is "a choice", for
+            // structs it it a correctness requirement (see "fgValueNumberBlockAssignment").
+            if (!indir->OperIs(GT_FIELD))
+            {
+                CORINFO_CLASS_HANDLE fieldClassHandle = NO_CLASS_HANDLE;
+                var_types            fieldType =
+                    m_compiler->eeGetFieldType(fieldSeq->GetTail()->GetFieldHandle(), &fieldClassHandle);
+
+                if ((fieldType != indir->TypeGet()) ||
+                    ((fieldType == TYP_STRUCT) && (fieldClassHandle != indirLayout->GetClassHandle())))
```

Comment:
Is `indirLayout` guaranteed to be non-null here?

Reply:
(Not sure why my reply was lost)

Yes, it is always set by `SelectLocalIndirTransform` when it returns something other than `IndirTransform::None`.

Reply:
> (Not sure why my reply was lost)

GitHub double posted my two comments so I deleted one of each, probably I managed to delete the one right as you posted.

> it is always set by SelectLocalIndirTransform when it returns something other than IndirTransform::None.

`SelectLocalIndirTransform` does not look to set it in the early-out `!varTypeIsStruct(indir)` case, which is why I was wondering.

Reply:
> `SelectLocalIndirTransform` does not look to set it in the early-out !varTypeIsStruct(indir) case, which is why I was wondering.

I see. It is a bit tricky, but should be correct:
```c
if ((fieldType != indir->TypeGet()) || // 'true`  -> we don't access "indirLayout" due to short-circuiting
                                       // 'false' -> "fieldType == indir->TypeGet()", so we access "indirLayout"
                                       // only if both are TYP_STRUCT.
    ((fieldType == TYP_STRUCT) && (fieldClassHandle != indirLayout->GetClassHandle())))
```
That said, I think we can just delete this code. It is not needed with physical VN.